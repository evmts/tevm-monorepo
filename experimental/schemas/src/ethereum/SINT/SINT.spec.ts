import {
	isINT8,
	isINT16,
	isINT32,
	isINT64,
	isINT128,
	isINT256,
	parseInt8,
	parseInt16,
	parseInt32,
	parseInt64,
	parseInt128,
	parseInt256,
} from './index.js'
import { assertType, describe, expect, it } from 'vitest'

const testCases = [
	{
		type: 'INT8',
		max: 127,
		min: -128,
		parse: parseInt8,
		isEqual: isINT8,
		exceedsMax: 128,
		belowMin: -129,
	},
	{
		type: 'INT16',
		max: 32767,
		min: -32768,
		parse: parseInt16,
		isEqual: isINT16,
		exceedsMax: 32768,
		belowMin: -32769,
	},
	{
		type: 'INT32',
		max: '2147483647',
		min: '-2147483648',
		parse: parseInt32,
		isEqual: isINT32,
		exceedsMax: '2147483648',
		belowMin: '-2147483649',
	},
	{
		type: 'INT64',
		max: '9223372036854775807',
		min: '-9223372036854775808',
		parse: parseInt64,
		isEqual: isINT64,
		exceedsMax: '9223372036854775808',
		belowMin: '-9223372036854775809',
	},
	{
		type: 'INT128',
		max: '170141183460469231731687303715884105727',
		min: '-170141183460469231731687303715884105728',
		parse: parseInt128,
		isEqual: isINT128,
		exceedsMax: '170141183460469231731687303715884105728',
		belowMin: '-170141183460469231731687303715884105729',
	},
	{
		type: 'INT256',
		max: BigInt(
			'0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF',
		).toString(),
		min: (-BigInt(
			'0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF',
		)).toString(),
		parse: parseInt256,
		isEqual: isINT256,
		exceedsMax: (
			BigInt(
				'0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF',
			) + BigInt(1)
		).toString(),
		belowMin: (
			-BigInt(
				'0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF',
			) - BigInt(1)
		).toString(),
	},
]

describe.each(testCases)('%j', (testCase) => {
	it(`should return true for valid ${testCase.type} values`, () => {
		expect(testCase.isEqual(BigInt(0))).toBe(true)
		expect(testCase.isEqual(BigInt(testCase.max))).toBe(true)
		expect(testCase.isEqual(BigInt(testCase.min))).toBe(true)
	})

	it(`should return false for out of bounds values for ${testCase.type}`, () => {
		expect(testCase.isEqual(BigInt(testCase.exceedsMax))).toBe(false)
		expect(testCase.isEqual(BigInt(testCase.belowMin))).toBe(false)
	})

	it(`should return false for non-bigint values for ${testCase.type}`, () => {
		expect(testCase.isEqual('0x52' as any)).toBe(false)
		expect(testCase.isEqual(testCase.max.toString() as any)).toBe(false)
		expect(testCase.isEqual('' as any)).toBe(false)
		expect(testCase.isEqual(true as any)).toBe(false)
		expect(testCase.isEqual({} as any)).toBe(false)
		expect(testCase.isEqual('not an int' as any)).toBe(false)
		expect(testCase.isEqual(Number(testCase.max) as any)).toBe(false)
		expect(testCase.isEqual(undefined as any)).toBe(false)
		expect(testCase.isEqual(null as any)).toBe(false)
	})
})
describe.each(testCases)('parse%j', (testCase) => {
	it(`should return a valid ${testCase.type}`, () => {
		const expected = BigInt(testCase.max)
		const result = testCase.parse(expected)
		expect(result).toMatchSnapshot(`"${testCase.max.toString()}"`)
		assertType<typeof expected>(result)
	})

	it(`should throw if value is out of positive ${testCase.type} range`, () => {
		expect(() =>
			testCase.parse(testCase.exceedsMax as any),
		).toThrowErrorMatchingSnapshot()
	})

	it(`should throw if value is out of negative ${testCase.type} range`, () => {
		expect(() =>
			testCase.parse(testCase.belowMin as any),
		).toThrowErrorMatchingSnapshot()
	})

	it(`should throw if value is not a bigint for ${testCase.type}`, () => {
		expect(() => testCase.parse(52 as any)).toThrowErrorMatchingSnapshot()
	})
})
