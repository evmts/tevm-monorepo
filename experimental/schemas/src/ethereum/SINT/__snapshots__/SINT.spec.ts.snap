// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse{"type":"INT8","max":127,"min":-128,"exceedsMax":128,"belowMin":-129} > should return a valid INT8 > "127" 1`] = `127n`;

exports[`parse{"type":"INT8","max":127,"min":-128,"exceedsMax":128,"belowMin":-129} > should throw if value is not a bigint for INT8 1`] = `
"InvalidINTError: Received 52 is too large to be an INT8. Must be < 2^7.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT8","max":127,"min":-128,"exceedsMax":128,"belowMin":-129} > should throw if value is out of negative INT8 range 1`] = `
"InvalidINTError: Received -129 is too small to be an INT8. Must be >= -2^7.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT8","max":127,"min":-128,"exceedsMax":128,"belowMin":-129} > should throw if value is out of positive INT8 range 1`] = `
"InvalidINTError: Received 128 is too large to be an INT8. Must be < 2^7.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT16","max":32767,"min":-32768,"exceedsMax":32768,"belowMin":-32769} > should return a valid INT16 > "32767" 1`] = `32767n`;

exports[`parse{"type":"INT16","max":32767,"min":-32768,"exceedsMax":32768,"belowMin":-32769} > should throw if value is not a bigint for INT16 1`] = `
"InvalidINTError: Received 52 is too large to be an INT16. Must be < 2^15.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT16","max":32767,"min":-32768,"exceedsMax":32768,"belowMin":-32769} > should throw if value is out of negative INT16 range 1`] = `
"InvalidINTError: Received -32769 is too small to be an INT16. Must be >= -2^15.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT16","max":32767,"min":-32768,"exceedsMax":32768,"belowMin":-32769} > should throw if value is out of positive INT16 range 1`] = `
"InvalidINTError: Received 32768 is too large to be an INT16. Must be < 2^15.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT32","max":"2147483647","min":"-2147483648","exceedsMax":"2147483648","belowMin":"-2147483649"} > should return a valid INT32 > "2147483647" 1`] = `2147483647n`;

exports[`parse{"type":"INT32","max":"2147483647","min":"-2147483648","exceedsMax":"2147483648","belowMin":"-2147483649"} > should throw if value is not a bigint for INT32 1`] = `
"InvalidINTError: Received 52 is too large to be an INT32. Must be < 2^31.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT32","max":"2147483647","min":"-2147483648","exceedsMax":"2147483648","belowMin":"-2147483649"} > should throw if value is out of negative INT32 range 1`] = `
"InvalidINTError: Received -2147483649 is too small to be an INT32. Must be >= -2^31.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT32","max":"2147483647","min":"-2147483648","exceedsMax":"2147483648","belowMin":"-2147483649"} > should throw if value is out of positive INT32 range 1`] = `
"InvalidINTError: Received 2147483648 is too large to be an INT32. Must be < 2^31.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT64","max":"9223372036854775807","min":"-9223372036854775808","exceedsMax":"9223372036854775808","belowMin":"-9223372036854775809"} > should return a valid INT64 > "9223372036854775807" 1`] = `9223372036854775807n`;

exports[`parse{"type":"INT64","max":"9223372036854775807","min":"-9223372036854775808","exceedsMax":"9223372036854775808","belowMin":"-9223372036854775809"} > should throw if value is not a bigint for INT64 1`] = `
"InvalidINTError: Received 52 is too large to be an INT64. Must be < 2^63.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT64","max":"9223372036854775807","min":"-9223372036854775808","exceedsMax":"9223372036854775808","belowMin":"-9223372036854775809"} > should throw if value is out of negative INT64 range 1`] = `
"InvalidINTError: Received -9223372036854775809 is too small to be an INT64. Must be >= -2^63.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT64","max":"9223372036854775807","min":"-9223372036854775808","exceedsMax":"9223372036854775808","belowMin":"-9223372036854775809"} > should throw if value is out of positive INT64 range 1`] = `
"InvalidINTError: Received 9223372036854775808 is too large to be an INT64. Must be < 2^63.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT128","max":"170141183460469231731687303715884105727","min":"-170141183460469231731687303715884105728","exceedsMax":"170141183460469231731687303715884105728","belowMin":"-170141183460469231731687303715884105729"} > should return a valid INT128 > "170141183460469231731687303715884105727" 1`] = `170141183460469231731687303715884105727n`;

exports[`parse{"type":"INT128","max":"170141183460469231731687303715884105727","min":"-170141183460469231731687303715884105728","exceedsMax":"170141183460469231731687303715884105728","belowMin":"-170141183460469231731687303715884105729"} > should throw if value is not a bigint for INT128 1`] = `
"InvalidINTError: Received 52 is too large to be an INT128. Must be < 2^127.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT128","max":"170141183460469231731687303715884105727","min":"-170141183460469231731687303715884105728","exceedsMax":"170141183460469231731687303715884105728","belowMin":"-170141183460469231731687303715884105729"} > should throw if value is out of negative INT128 range 1`] = `
"InvalidINTError: Received -170141183460469231731687303715884105729 is too small to be an INT128. Must be >= -2^127.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT128","max":"170141183460469231731687303715884105727","min":"-170141183460469231731687303715884105728","exceedsMax":"170141183460469231731687303715884105728","belowMin":"-170141183460469231731687303715884105729"} > should throw if value is out of positive INT128 range 1`] = `
"InvalidINTError: Received 170141183460469231731687303715884105728 is too large to be an INT128. Must be < 2^127.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT256","max":"57896044618658097711785492504343953926634992332820282019728792003956564819967","min":"-57896044618658097711785492504343953926634992332820282019728792003956564819967","exceedsMax":"57896044618658097711785492504343953926634992332820282019728792003956564819968","belowMin":"-57896044618658097711785492504343953926634992332820282019728792003956564819968"} > should return a valid INT256 > "57896044618658097711785492504343953926634992332820282019728792003956564819967" 1`] = `57896044618658097711785492504343953926634992332820282019728792003956564819967n`;

exports[`parse{"type":"INT256","max":"57896044618658097711785492504343953926634992332820282019728792003956564819967","min":"-57896044618658097711785492504343953926634992332820282019728792003956564819967","exceedsMax":"57896044618658097711785492504343953926634992332820282019728792003956564819968","belowMin":"-57896044618658097711785492504343953926634992332820282019728792003956564819968"} > should throw if value is not a bigint for INT256 1`] = `
"InvalidINTError: Received 52 is too large to be an INT256. Must be < 2^255.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT256","max":"57896044618658097711785492504343953926634992332820282019728792003956564819967","min":"-57896044618658097711785492504343953926634992332820282019728792003956564819967","exceedsMax":"57896044618658097711785492504343953926634992332820282019728792003956564819968","belowMin":"-57896044618658097711785492504343953926634992332820282019728792003956564819968"} > should throw if value is out of negative INT256 range 1`] = `
"InvalidINTError: Received -57896044618658097711785492504343953926634992332820282019728792003956564819968 is too large to be an INT256. Must be < 2^255.
https://tevm.sh/reference/errors"
`;

exports[`parse{"type":"INT256","max":"57896044618658097711785492504343953926634992332820282019728792003956564819967","min":"-57896044618658097711785492504343953926634992332820282019728792003956564819967","exceedsMax":"57896044618658097711785492504343953926634992332820282019728792003956564819968","belowMin":"-57896044618658097711785492504343953926634992332820282019728792003956564819968"} > should throw if value is out of positive INT256 range 1`] = `
"InvalidINTError: Received 57896044618658097711785492504343953926634992332820282019728792003956564819968 is too large to be an INT256. Must be < 2^255.
https://tevm.sh/reference/errors"
`;
